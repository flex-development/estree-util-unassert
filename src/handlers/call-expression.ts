/**
 * @file Handlers - CallExpression
 * @module estree-util-unassert/handlers/CallExpression
 */

import type { HandlerContext } from '#src/interfaces'
import type { Key } from '#src/types'
import {
  DOT,
  at,
  define,
  set,
  type Optional
} from '@flex-development/tutils'
import { ok, unreachable } from 'devlop'
import type * as estree from 'estree'
import { CONTINUE, type Index } from 'estree-util-visit'
import { is } from 'unist-util-is'

/**
 * Add an ancestor node to the trash to be removed on exit or replace `node` if
 * it contains an assertion identifier.
 *
 * Supported parents:
 *
 * - {@linkcode estree.ArrayExpression}
 * - {@linkcode estree.ArrowFunctionExpression}
 * - {@linkcode estree.AssignmentExpression}
 * - {@linkcode estree.AssignmentPattern}
 * - {@linkcode estree.AwaitExpression}
 * - {@linkcode estree.CallExpression}
 * - {@linkcode estree.ConditionalExpression}
 * - {@linkcode estree.ExportDefaultDeclaration}
 * - {@linkcode estree.ExpressionStatement}
 * - {@linkcode estree.LogicalExpression}
 * - {@linkcode estree.Property}
 * - {@linkcode estree.ReturnStatement}
 * - {@linkcode estree.UnaryExpression}
 * - {@linkcode estree.YieldExpression}
 *
 * @see {@linkcode HandlerContext}
 * @see {@linkcode estree.CallExpression}
 *
 * @this {HandlerContext}
 *
 * @param {estree.CallExpression} node - Node being entered
 * @param {Key} key - Field at which `node` lives in `this.parent` (or where a
 * list of nodes live if `this.parent[key]` is an array)
 * @param {Optional<Index>} index - Index where `node` lives in `this.parent`,
 * or `undefined` if `this.parent[key]` is not an array
 * @return {typeof CONTINUE} Next action
 */
function CallExpression(
  this: HandlerContext,
  node: estree.CallExpression,
  key: Key,
  index: Optional<Index>
): typeof CONTINUE {
  if (
    (
      is(node.callee, 'Identifier') &&
      this.identifiers.has(node.callee.name)
    ) ||
    (
      is(node.callee, 'MemberExpression') &&
      is(node.callee.object, 'Identifier') &&
      this.identifiers.has(node.callee.object.name)
    ) ||
    (
      is(node.callee, 'MemberExpression') &&
      is(node.callee.object, 'Identifier') &&
      is(node.callee.property, 'Identifier') &&
      node.callee.object.name === 'console' &&
      node.callee.property.name === 'assert'
    )
  ) {
    ok(this.parent, 'expected `parent`')
    ok(key, 'expected `key`')
    ok(key in this.parent, `expected \`parent.${key}\``)

    /**
     * Zero (`0`) node.
     *
     * @const {estree.Literal} zero
     */
    const zero: estree.Literal = { raw: '0', type: 'Literal', value: 0 }

    /**
     * Void zero (`void 0`) expression node.
     *
     * @const {estree.UnaryExpression} void0
     */
    const void0: estree.UnaryExpression = {
      argument: zero,
      operator: 'void',
      prefix: true,
      type: 'UnaryExpression'
    }

    // add ancestor to trash or replace node
    switch (this.parent.type) {
      case 'ArrayExpression':
      case 'CallExpression':
        ok(typeof index === 'number', 'expected `index` to be a number')
        set(this.parent, key + DOT + index, void0)
        break
      case 'AssignmentExpression':
        if (is(this.grandparent, 'ExpressionStatement')) {
          this.trash.add(this.grandparent)
        } else {
          define(this.parent, key, { value: void0 })
        }
        break
      case 'ArrowFunctionExpression':
      case 'AssignmentPattern':
      case 'ConditionalExpression':
      case 'LogicalExpression':
      case 'Property':
        define(this.parent, key, { value: void0 })
        break
      case 'AwaitExpression':
        if (is(this.grandparent, 'ExpressionStatement')) {
          this.trash.add(this.grandparent)
        } else {
          define(this.parent, key, { value: void0 })
        }
        break
      case 'ExpressionStatement':
        this.trash.add(this.parent)

        if (is(this.grandparent, 'IfStatement')) {
          !this.grandparent.alternate && this.trash.add(this.grandparent)
        }
        break
      case 'ExportDefaultDeclaration':
        define(this.parent, key, { value: at(node.arguments, 0, void0) })
        break
      case 'ReturnStatement':
      case 'YieldExpression':
        define(this.parent, key, { value: null })
        break
      case 'UnaryExpression':
        if (is(this.grandparent, 'ExpressionStatement')) {
          this.trash.add(this.grandparent)
        } else {
          define(this.parent, key, {
            value: this.parent.operator === 'void' ? zero : void0
          })
        }
        break
      /* c8 ignore next 3 */
      default:
        console.dir(this.parent, { depth: 10 })
        void unreachable(`unexpected parent: ${this.parent.type}`)
    }
  }

  return CONTINUE
}

export default CallExpression
