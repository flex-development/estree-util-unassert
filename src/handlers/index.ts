/**
 * @file Entry Point - Handlers
 * @module estree-util-unassert/handlers
 */

export { default as AssignmentExpression } from './assignment-expression'
