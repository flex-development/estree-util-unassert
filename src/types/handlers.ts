/**
 * @file Type Definitions - Handlers
 * @module estree-util-unassert/types/Handlers
 */

import type { Handler } from '#src/interfaces'
import type { Node } from 'estree'

/**
 * Node handlers map.
 */
type Handlers = { [K in Node['type']]?: Handler<Extract<Node, { type: K }>> }

export type { Handlers as default }
