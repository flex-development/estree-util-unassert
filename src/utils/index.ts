/**
 * @file Entry Point - Utilities
 * @module estree-util-unassert/utils
 */

export { default as getIdentifiers } from './get-identifiers'
export { default as hasModuleId } from './has-module-id'
export { default as MODULES_REGEX } from './modules-regex'
