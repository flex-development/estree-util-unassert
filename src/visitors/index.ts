/**
 * @file Entry Point - Visitors
 * @module estree-util-unassert/visitors
 */

export { default as enter } from './enter'
export { default as leave } from './leave'
