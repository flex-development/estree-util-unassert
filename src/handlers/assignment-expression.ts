/**
 * @file Handlers - AssignmentExpression
 * @module estree-util-unassert/handlers/AssignmentExpression
 */

import type { HandlerContext } from '#src/interfaces'
import { getIdentifiers, hasModuleId } from '#src/utils'
import type * as estree from 'estree'
import { CONTINUE } from 'estree-util-visit'
import { is } from 'unist-util-is'

/**
 * Collect assertion module identifiers from assignment expressions where the
 * righthand side (`node.right`) is a dynamic import or `require` call.
 *
 * If `node` contains an assertion module id, and its parent (`this.parent`) is
 * an {@linkcode estree.ExpressionStatement}, its parent will be added to the
 * trash to be removed on exit.
 *
 * Supported righthands (`node.right`):
 *
 * - {@linkcode estree.AwaitExpression}
 * - {@linkcode estree.CallExpression}
 *
 * @see {@linkcode HandlerContext}
 * @see {@linkcode estree.AssignmentExpression}
 *
 * @this {HandlerContext}
 *
 * @param {estree.AssignmentExpression} node - Node being entered
 * @return {typeof CONTINUE} Next action
 */
function AssignmentExpression(
  this: HandlerContext,
  node: estree.AssignmentExpression
): typeof CONTINUE {
  if (
    node.operator === '=' &&
    hasModuleId(node.right, this.modules) &&
    is(this.parent, 'ExpressionStatement')
  ) {
    this.trash.add(this.parent)
    getIdentifiers(node, this.identifiers)
  }

  return CONTINUE
}

export default AssignmentExpression
