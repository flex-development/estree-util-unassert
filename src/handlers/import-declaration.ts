/**
 * @file Handlers - ImportDeclaration
 * @module estree-util-unassert/handlers/ImportDeclaration
 */

import type { HandlerContext } from '#src/interfaces'
import { getIdentifiers, hasModuleId } from '#src/utils'
import type { Keys, Optional } from '@flex-development/tutils'
import { ok } from 'devlop'
import type * as estree from 'estree'
import { SKIP } from 'estree-util-visit'

/**
 * Collect assertion module identifiers when entering `node`.
 *
 * If `node` represents an imported assertion module, it will be added to the
 * trash to be removed on exit.
 *
 * @see {@linkcode HandlerContext}
 * @see {@linkcode estree.ImportDeclaration}
 *
 * @this {HandlerContext}
 *
 * @param {estree.ImportDeclaration} node - Node being entered
 * @param {Keys<estree.Node>[number]} key - Field at which `node` lives in
 * `this.parent` (or where a list of nodes lives)
 * @param {Optional<number>} index - Index where `node` lives, or `undefined`
 * if `this.parent[key]` is not an array
 * @return {[typeof SKIP, number]} Visitor action
 */
function ImportDeclaration(
  this: HandlerContext,
  node: estree.ImportDeclaration,
  key: Keys<estree.Node>[number],
  index: Optional<number>
): [typeof SKIP, number] {
  ok(typeof index === 'number', 'expected `index` to be a number')
  ok(!Number.isNaN(index), 'expected `index` not be NaN')

  // trash node and collect assertion module identifiers
  if (hasModuleId(node, this.modules)) {
    this.trash.add(node)
    getIdentifiers(node, this.identifiers)
  }

  return [SKIP, index + 1]
}

export default ImportDeclaration
