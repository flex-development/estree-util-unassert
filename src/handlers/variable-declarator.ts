/**
 * @file Handlers - VariableDeclarator
 * @module estree-util-unassert/handlers/VariableDeclarator
 */

import type { HandlerContext } from '#src/interfaces'
import { getIdentifiers, hasModuleId } from '#src/utils'
import type * as estree from 'estree'
import { CONTINUE } from 'estree-util-visit'
import { is } from 'unist-util-is'

/**
 * Collect assertion module identifiers from variable declarators where the
 * initializer (`node.init`) is a dynamic import or `require` call.
 *
 * If `node` is the child of a `VariableDeclaration` containing one declaration,
 * its parent (`this.parent`) will be added to the trash to be removed on exit.
 * Otherwise, `node` will be added to the trash.
 *
 * Supported initializers (`node.init`):
 *
 * - {@linkcode estree.AwaitExpression}
 * - {@linkcode estree.CallExpression}
 *
 * @see {@linkcode HandlerContext}
 * @see {@linkcode estree.VariableDeclarator}
 *
 * @this {HandlerContext}
 *
 * @param {estree.VariableDeclarator} node - Node being entered
 * @return {typeof CONTINUE} Visitor action
 */
function VariableDeclarator(
  this: HandlerContext,
  node: estree.VariableDeclarator
): typeof CONTINUE {
  if (
    hasModuleId(node.init, this.modules) &&
    is(this.parent, 'VariableDeclaration')
  ) {
    this.trash.add(this.parent.declarations.length === 1 ? this.parent : node)
    getIdentifiers(node.id, this.identifiers)
  }

  return CONTINUE
}

export default VariableDeclarator
