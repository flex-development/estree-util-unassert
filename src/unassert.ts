/**
 * @file unassert
 * @module estree-util-unassert/unassert
 */

import type { Nilable } from '@flex-development/tutils'
import { createFilter } from '@rollup/pluginutils'
import type { Node, Program } from 'estree'
import { visit } from 'estree-util-visit'
import * as handlers from './handlers'
import type { HandlerContext, Options } from './interfaces'
import { MODULES_REGEX } from './utils'
import * as visitors from './visitors'

/**
 * Remove assertions.
 *
 * @see {@linkcode Options}
 * @see {@linkcode Program}
 *
 * @this {void}
 *
 * @param {Program} tree - JavaScript syntax tree
 * @param {Nilable<Options>?} [options] - Configuration options
 * @return {void} Nothing
 */
function unassert(this: void, tree: Program, options?: Nilable<Options>): void {
  options ??= {}
  options.modules ??= MODULES_REGEX

  /**
   * Node handler context.
   *
   * @const {HandlerContext} context
   */
  const context: HandlerContext = {
    grandparent: null,
    identifiers: new Set<string>(),
    modules: createFilter(options.modules, [], { resolve: false }),
    parent: null,
    trash: new WeakSet<Node>()
  }

  return void visit(tree, {
    enter: visitors.enter(context, handlers),
    leave: visitors.leave(context)
  })
}

export default unassert
