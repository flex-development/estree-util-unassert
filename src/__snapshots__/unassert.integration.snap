// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`integration:unassert > remove assertions > program sample 0: add.mjs 1`] = `
"function add(a, b) {
  return a + b;
}
export default add;
"
`;

exports[`integration:unassert > remove assertions > program sample 1: assert.mjs 1`] = `
"function assert(value, message, ...params) {
  return void 0;
}
export default assert;
"
`;

exports[`integration:unassert > remove assertions > program sample 2: call-expression.mjs 1`] = `
"import {DOT, at, define, set} from "@flex-development/tutils";
import {CONTINUE} from "estree-util-visit";
import {is} from "unist-util-is";
function CallExpression(node, key, index) {
  if (is(node.callee, "Identifier") && this.identifiers.has(node.callee.name) || is(node.callee, "MemberExpression") && is(node.callee.object, "Identifier") && this.identifiers.has(node.callee.object.name) || is(node.callee, "MemberExpression") && is(node.callee.object, "Identifier") && is(node.callee.property, "Identifier") && node.callee.object.name === "console" && node.callee.property.name === "assert") {
    const zero = {
      raw: "0",
      type: "Literal",
      value: 0
    };
    const void0 = {
      argument: zero,
      operator: "void",
      prefix: true,
      type: "UnaryExpression"
    };
    switch (this.parent.type) {
      case "ArrayExpression":
      case "CallExpression":
        set(this.parent, key + DOT + index, void0);
        break;
      case "AssignmentExpression":
        if (is(this.grandparent, "ExpressionStatement")) {
          this.trash.add(this.grandparent);
        } else {
          define(this.parent, key, {
            value: void0
          });
        }
        break;
      case "ArrowFunctionExpression":
      case "AssignmentPattern":
      case "ConditionalExpression":
      case "LogicalExpression":
      case "Property":
        define(this.parent, key, {
          value: void0
        });
        break;
      case "AwaitExpression":
        if (is(this.grandparent, "ExpressionStatement")) {
          this.trash.add(this.grandparent);
        } else {
          define(this.parent, key, {
            value: void0
          });
        }
        break;
      case "ExpressionStatement":
        this.trash.add(this.parent);
        break;
      case "ExportDefaultDeclaration":
        define(this.parent, key, {
          value: at(node.arguments, 0, void0)
        });
        break;
      case "ReturnStatement":
      case "YieldExpression":
        define(this.parent, key, {
          value: null
        });
        break;
      case "UnaryExpression":
        if (is(this.grandparent, "ExpressionStatement")) {
          this.trash.add(this.grandparent);
        } else {
          define(this.parent, key, {
            value: this.parent.operator === "void" ? zero : void0
          });
        }
        break;
      default:
        console.dir(this.parent, {
          depth: 10
        });
    }
  }
  return CONTINUE;
}
var call_expression_default = CallExpression;
export {call_expression_default as default};
"
`;

exports[`integration:unassert > remove assertions > program sample 3: century-from-year.mjs 1`] = `
"const centuryFromYear = year => {
  return Math.ceil(year / 100);
};
export default centuryFromYear;
"
`;

exports[`integration:unassert > remove assertions > program sample 4: digitize.mjs 1`] = `
"function digitize(n) {
  if (n <= 9) return [n];
  const digits = [];
  while (n > 0) digits.push(n % 10 | 0) && (n = n / 10 | 0);
  return digits;
}
export default digitize;
"
`;

exports[`integration:unassert > remove assertions > program sample 5: divide.cjs 1`] = `
""use strict";
let bar = "BAR", foo = "FOO";
function divide(a, b) {
  return a / b;
}
module.exports = module.exports.default = divide;
"
`;

exports[`integration:unassert > remove assertions > program sample 6: gemoji-html.mjs 1`] = `
"import {nameToEmoji} from "gemoji";
import {codes} from "micromark-util-symbol";
function gemojiHtml() {
  return {
    enter: {
      gemoji() {
        return void this.tag("<span>");
      }
    },
    exit: {
      gemoji(token) {
        const val = this.sliceSerialize(token);
        this.raw(nameToEmoji[val.slice(1, -1)] ?? val);
        return void this.tag("</span>");
      }
    }
  };
}
var html_default = gemojiHtml;
export {html_default as default};
"
`;

exports[`integration:unassert > remove assertions > program sample 7: gemoji-shortcode.mjs 1`] = `
"import {asciiAlphanumeric} from "micromark-util-character";
import {codes} from "micromark-util-symbol";
function previous(code) {
  return code !== codes.backslash && code !== codes.colon;
}
var shortcode_default = {
  name: "gemoji",
  previous,
  tokenize(effects, ok, nok) {
    function inside(code) {
      switch (true) {
        case code === codes.colon:
          effects.consume(code);
          effects.exit("gemoji");
          return ok;
        case asciiAlphanumeric(code):
        case code === codes.dash:
        case code === codes.plusSign:
        case code === codes.underscore:
          effects.consume(code);
          return inside;
        default:
          return nok(code);
      }
    }
    function begin(code) {
      switch (code) {
        case codes.eof:
        case codes.colon:
          return nok(code);
        default:
          effects.consume(code);
          return inside;
      }
    }
    const start = code => {
      effects.enter("gemoji");
      effects.consume(code);
      return begin;
    };
    return start;
  }
};
export {shortcode_default as default};
"
`;

exports[`integration:unassert > remove assertions > program sample 8: http-assert.cjs 1`] = `
""use strict";
module.exports = module.exports.default = function (username, check) {
  try {} catch (err) {}
};
"
`;

exports[`integration:unassert > remove assertions > program sample 9: human-readable.mjs 1`] = `
"async function humanReadable(seconds) {
  return new Promise(resolve => {
    let formatted = "";
    for (const converter of [3600, 60, 1]) {
      const time = seconds / converter | 0;
      formatted += time < 10 ? \`0\${time}\` : time;
      if (converter !== 1) formatted += ":";
      seconds -= time * converter;
    }
    return resolve(formatted);
  });
}
export default humanReadable;
"
`;

exports[`integration:unassert > remove assertions > program sample 10: multiply.cjs 1`] = `
""use strict";
const multiply = function (a, b) {
  return a * b;
};
module.exports = module.exports.default = multiply;
"
`;

exports[`integration:unassert > remove assertions > program sample 11: noop.mjs 1`] = `
"export default () => void 0;
"
`;

exports[`integration:unassert > remove assertions > program sample 12: subtract.mjs 1`] = `
"function subtract(a, b) {
  return a - b;
}
export default subtract;
"
`;
